pkg/jetstream/publisher.go

```diff
@@ -0,0 +1,78 @@
+package jetstream
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/ThreeDotsLabs/watermill"
+	wmnats "github.com/ThreeDotsLabs/watermill-nats/v2/pkg/nats"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+)
+
+var _ message.Publisher = &Publisher{}
+
+// Publisher provides a watermill publisher interface to NATS JetStream
+type Publisher struct {
+	nc     *nats.Conn
+	js     jetstream.JetStream
+	m      wmnats.Marshaler
+	logger watermill.LoggerAdapter
+	known  sync.Map
+}
+
+// NewPublisher creates a new watermill JetStream publisher.
+func NewPublisher(config *PublisherConfig) (*Publisher, error) {
+	config.setDefaults()
+
+	nc, err := nats.Connect(config.URL)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect: %w", err)
+	}
+
+	return newPublisher(nc, config)
+}
+
+func newPublisher(nc *nats.Conn, config *PublisherConfig) (*Publisher, error) {
+	jsapi, err := jetstream.New(nc)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize jetstream: %w", err)
+	}
+
+	return &Publisher{
+		nc:     nc,
+		js:     jsapi,
+		m:      &wmnats.NATSMarshaler{},
+		logger: config.Logger,
+		known:  sync.Map{},
+	}, nil
+}
+
+// Publish sends provided watermill messages to the given topic.
+func (p *Publisher) Publish(topic string, messages ...*message.Message) error {
```

Comment:
Decided to draw a hard line here and only allow auto provisioning by subscribers.  Should keep things simpler I think.